proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,159 @@
+
+This proposal introduces secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective APIs. This would allow for more secure and dynamic secret retrieval, eliminating the need to store secrets in the filesystem and reducing the potential for unauthorized access.
+
+### Pitfalls of the current solution
+
+Storing secrets in the filesystem poses risks, especially in environments like Kubernetes, where any pod on a node can access files mounted on that node. This could expose secrets to attackers. Additionally, configuring secrets through the filesystem often requires extra setup steps in some environments, which can be cumbersome for users.
+
+Storing secrets inline can also pose risks, as the configuration file may still be accessible through the filesystem. Additionally it can lead to configuration files becoming cluttered and difficult to manage.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow Prometheus to read secrets remotely from secret providers.
+  * Anywhere in the configs([1](https://prometheus.io/docs/prometheus/latest/configuration/configuration/),[2](https://prometheus.io/docs/alerting/latest/configuration/#configuration-file-introduction)) with the `<secret>` type, it should be possible to fetch from secret providers
+  * This will include Alertmanager 
+* Introduce secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective API.
+* Backwards compatibility for secrets in config
+
+### Audience
+
+* Prometheus maintainers
+* Alertmanager maintainers
+* Secret providers interested in contributing code
+* Users looking to use secret providers
+
+## Non-Goals
+
+* Support for non-string secret values
+* Secret transformations/processing
+  * Things like String concatenation, base64 encoding/decoding
+* Default values
+* De-duplication and re-using common config values for secret provider configs
+  * This will be left to a follow up proposal if needed
+
+## How
+
+
+### Configuration
+
+Wherever a `<secret>` type is present in the configuration files, we will allow users to specify a map specifying how to fetch from a secret provider. The generic format would be
+
+```
+secret_field:
+  provider:  <type of the provider>
+  <property1>: <value1>
+  ...
+  <propertyN>: <valueN>
+```
+
+For example when specifying a password fetched from the kubernetes provider with an id of `pass2` in namespace `ns1` for the HTTP passsword field it would look like this:
+
+```
+        password:
+          provider: kubernetes
+          namespace: ns1
+          secret_id: pass2
```

Comment:
What's secret ID? Maybe better to make it obvious e.g.		

```suggestion
          provider: kubernetes
          namespace: <ns>
	  name: <secret name>
          key: <data's key for secret name>
```

Reply:
Thanks! accepted it